### ISSUE TITLE
RMSProp does not match original Tensorflow impl

### ISSUE THREAD

**Issue by rwightman** [deepmind/optax#49]
A lot of papers using RMSProp have hparams that were tuned with the original Tensorflow impl. 

The Optax impl is missing the momentum option and initializes the rms value differently.

TF1 RMSProp (https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/training/rmsprop.py#L126)
* has momentum option
* initializes RMS value to ones

Keras RMSProp (https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/optimizer_v2/rmsprop.py#L35-L299)
* has momentum option
* initializes RMS values to zeros (as with here and PyTorch)

I've spent time replicating results of papers like EfficientNet (and related) in PyTorch and ended up using my own RMSProp impl that matches the TF1 variant (the PyTorch one does not either).

**Comment by 8bitmp3** [comment]
@rwightman Sonnet (for TensorFlow) also has a momentum arg if I'm not mistaken https://github.com/deepmind/sonnet/blob/5a465696383f967d5bffb6599347d9e6c15cef4b/sonnet/src/optimizers/rmsprop.py 🤔 (though it's "not Nesterov momentum")

**Comment by 8bitmp3** [comment]
To provide more context, here's a code comparison:

Optax https://github.com/deepmind/optax/blob/70e0dd04c7bad9359b372260c56cd85c731b3cbe/optax/_src/alias.py

```py
def rmsprop(learning_rate: ScalarOrSchedule,
            decay: float = 0.9,
            eps: float = 1e-8,
            centered: bool = False) -> GradientTransformation:
  if centered:
    return combine.chain(
        transform.scale_by_stddev(decay=decay, eps=eps),
        _scale_by_learning_rate(learning_rate),
    )
  return combine.chain(
      transform.scale_by_rms(decay=decay, eps=eps),
      _scale_by_learning_rate(learning_rate),
  )
```

Sonnet https://github.com/deepmind/sonnet/blob/5a465696383f967d5bffb6599347d9e6c15cef4b/sonnet/src/optimizers/rmsprop.py

```py
def rmsprop_update(update, decay, learning_rate, epsilon, mu, mom, ms, mg):
  """Computes a single RMSProp update."""
  ms = tf.square(update) * (1. - decay) + ms * decay
  if mg is not None:  # centered
    mg = update * (1. - decay) + mg * decay
    denominator = ms - tf.square(mg) + epsilon
  else:
    denominator = ms + epsilon
  mom = (mu * mom) + (learning_rate * update * tf.math.rsqrt(denominator))
  return mom, ms, mg

class RMSProp(base.Optimizer):
  """RMSProp module.
  ...
  Attributes:
    learning_rate: Learning rate.
    decay: Learning rate decay over each update.
    momentum: Momentum scalar.
    epsilon: Small value to avoid zero denominator.
    centered: `True` if centered.
    mom: Accumulated mom for each parameter.
    ms: Accumulated ms for each parameter.
    mg: Accumulated mg for each parameter.
  """
...
```

**Comment by 8bitmp3** [comment]
I wonder if (Nesterov) momentum is something you currently add outside RMSprop 🤷‍♂️ In `/optax/optax/_src/transform.py`

https://github.com/deepmind/optax/blob/01e846a4ca66aa1118428d152cf97ad7b7acaed9/optax/_src/transform.py

```py
"""Gradient transformations."""
```

```py
class TraceState(OptState):
  """Holds an aggregation of past updates."""
  trace: Params


def trace(decay: float, nesterov: bool) -> GradientTransformation:
  """Compute a trace of past updates.
  Args:
    decay: the decay rate for the tracing of past updates.
    nesterov: whether to use Nesterov momentum.
  Returns:
    An (init_fn, update_fn) tuple.
  """

  ...
```

Sorry I can't read code well 😄

**Comment by mtthss** [comment]
> A lot of papers using RMSProp have hparams that were tuned with the original Tensorflow impl.

While it's true that TF1 optims initialised the stats with zero it seems like the community has moved in the opposite direction: for instance optax's initialisation with zeros matches both `Keras`, `PyTorch`, `flax.optim`, `jax.experimental.optimizers`. So I would be weary to overfit to the TF1 initialisation. What we could do is to have an `initial_scale` with default 0, that you can use to initialise stats with ones as in TF1 if needed.

Would that address your needs?

> Regarding momentum

You can easily compose optax's gradient transformation into custom optimisers, by chaining for instance `scale_by_rms` and `trace`

**Comment by rwightman** [comment]
@mtthss yes, an initial_scale arg could achieve my goals

I had taken a look and it appears the snippet below would match RMSProp w/ momentum. I imagine someone else will request momentum as a n out of the box capability at some point as it's fairly common to have it enabled.

One thing to mention re the centered variant, there are two scale params, in TF1 one is init to zero and the other is one. I don't really care as much about this case since I never use it. Ignore, or `initial_scale: Union[float, tuple[Float]]`? 

```
def rmsprop_momentum(
    learning_rate: ScalarOrSchedule,
    decay: float = 0.9,
    momentum: float = 0,
    eps: float = 1e-8,
    centered: bool = False,
    initial_scale: float = 1.) -> GradientTransformation:
  if centered:
    return combine.chain(
        transform.scale_by_stddev(decay=decay, eps=eps),
        _scale_by_learning_rate(learning_rate),
        transform.trace(decay=momentum, nesterov=False),
    )
  return combine.chain(
      transform.scale_by_rms(decay=decay, eps=eps, initial_scale=initial_scale),
      _scale_by_learning_rate(learning_rate),
      transform.trace(decay=momentum, nesterov=False),
  )
```

**Comment by mtthss** [comment]
The accumulators of rmsprop/adam have one the semantic of a `mean` and the of a scale` factor
I think everyone sets the mean one to 0, its only the `scale` one that is more controversial (either 0 or 1 depending on the framerwork) so I would suggest `initial_scale` to only refer to that one (i.e. be just a float), the name also then corresponds well to its semantics

Your code looks good,
want to put together a PR adding this option?

**Comment by rwightman** [comment]
@mtthss valid point, I have some variants of this testing right now, can create a PR. 

Would you rather the PR be limited only to `initial_scale` arg addition, or also include the inclusion of trace to the `rmsprop` alias. With #52/#54 there wouldn't be any overhead when momentum defaults to 0.

There are a lot of popular models trained with rmsprop using momentum=0.9. It was/is a default setting for the TF Slim CNN training scripts, and also many of the TF TPU example scripts. So models like EfficientNet, MNasNet, MobileNetV3/V2, NASNet, PNASNet, AmoebaNet, InceptionV2/V3/V4, etc all trained with momentum=0.9.

**Comment by mtthss** [comment]
I think the PR can include both

**Comment by rwightman** [comment]
@mtthss Since I have this already for my testing, pre PR review. Aside from my formatting changes which will not be included in PR, any issues with

initial_scale changes for rms/stddev:
https://github.com/rwightman/efficientnet-jax/blob/optax/jeffnet/common/optim/rmsprop.py#L8-L64

rmsprop alias update:
https://github.com/rwightman/efficientnet-jax/blob/optax/jeffnet/common/optim/rmsprop.py#L88-L106
